p2p/memp2p/memp2p_network.go

```diff
@@ -0,0 +1,159 @@
+	if resultingLength <= 0 {
+		network.mutex.Unlock()
+		return []string{}
+	}
+	addresses := make([]string, resultingLength)
+	k := 0
+	for _, peerID := range network.peerIDs {
+		if peerID == peerIDToExclude {
+			continue
+		}
+		addresses[k] = fmt.Sprintf("/memp2p/%s", peerID)
+		k++
+	}
+	network.mutex.Unlock()
+	return addresses
+}
+
+// Peers provides a copy of its internal map of peers
+func (network *MemP2PNetwork) Peers() map[p2p.PeerID]*MemP2PMessenger {
+	network.mutex.RLock()
+	peersCopy := make(map[p2p.PeerID]*MemP2PMessenger)
+	for peerID, peer := range network.peers {
+		peersCopy[peerID] = peer
+	}
+	network.mutex.RUnlock()
+	return peersCopy
+}
+
+// PeersExceptOne provides a copy of its internal map of peers, excluding a specific peer.
+func (network *MemP2PNetwork) PeersExceptOne(peerIDToExclude p2p.PeerID) map[p2p.PeerID]*MemP2PMessenger {
+	network.mutex.RLock()
+	peersCopy := make(map[p2p.PeerID]*MemP2PMessenger)
+	for peerID, peer := range network.peers {
+		if peerID == peerIDToExclude {
+			continue
+		}
+		peersCopy[peerID] = peer
+	}
+	network.mutex.RUnlock()
+	return peersCopy
+}
+
+// PeerIDs provides a copy of its internal slice of peerIDs
+func (network *MemP2PNetwork) PeerIDs() []p2p.PeerID {
+	network.mutex.RLock()
+	peerIDsCopy := make([]p2p.PeerID, len(network.peerIDs))
+	for i, peer := range network.peerIDs {
+		peerIDsCopy[i] = peer
+	}
+	network.mutex.RUnlock()
+	return peerIDsCopy
+}
+
+//PeerIDsExceptOne provides a copy of its internal slice of peerIDs, excluding a specific peer.
+func (network *MemP2PNetwork) PeerIDsExceptOne(peerIDToExclude p2p.PeerID) []p2p.PeerID {
+	network.mutex.RLock()
+	resultingLength := len(network.peerIDs) - 1
+	if resultingLength <= 0 {
+		network.mutex.RUnlock()
+		return []p2p.PeerID{}
+	}
+
+	peerIDsCopy := make([]p2p.PeerID, resultingLength)
+	k := 0
+	for _, peerID := range network.peerIDs {
+		if peerID == peerIDToExclude {
+			continue
+		}
+		peerIDsCopy[k] = peerID
+		k++
+	}
+	network.mutex.RUnlock()
+	return peerIDsCopy
+}
+
+// RegisterPeer adds a messenger to the Peers map and its PeerID to the peerIDs
+// slice.
+func (network *MemP2PNetwork) RegisterPeer(messenger *MemP2PMessenger) {
+	network.mutex.RLock()
+	network.peerIDs = append(network.peerIDs, messenger.ID())
+	network.peers[messenger.ID()] = messenger
+	network.mutex.RUnlock()
+}
+
+// UnregisterPeer removes a messenger from the Peers map and its PeerID from
+// the peerIDs slice.
+func (network *MemP2PNetwork) UnregisterPeer(peerID p2p.PeerID) {
+	network.mutex.RLock()
+	// Delete from the Peers map.
+	delete(network.peers, peerID)
+	// Remove from the peerIDs slice, maintaining the order of the slice.
+	index := -1
+	for i, id := range network.peerIDs {
+		if id == peerID {
+			index = i
+		}
+	}
+	network.peerIDs = append(network.peerIDs[0:index], network.peerIDs[index+1:]...)
+	network.mutex.RUnlock()
+}
+
+func (network *MemP2PNetwork) LogMessage(message p2p.MessageP2P) {
+	network.mutex.RLock()
```

Comment:
I would prefer separate mutexes for separate resources.
Also it needs to be Lock/Unlock instead.

Reply:
Added a separate mutex for the message log. Also replaced with `Lock`/`Unlock`.